clients/venice-thin-client/src/main/java/com/linkedin/venice/client/store/AbstractAvroStoreClient.java

```diff
@@ -272,28 +272,36 @@ protected RecordSerializer<K> getKeySerializerForRequest() {
     }
     synchronized (this) {
       try {
+        if (keySerializer != null) {
+          whetherStoreInitTriggeredByRequestFail = false;
+          return keySerializer;
+        }
         return getKeySerializerWithRetryWithShortInterval();
       } catch (Exception e) {
         whetherStoreInitTriggeredByRequestFail = true;
         // Kick off an async thread to keep retrying
-        asyncStoreInitThread = new Thread(() -> {
-          while (true) {
-            try {
-              getKeySerializerWithRetryWithShortInterval();
-              whetherStoreInitTriggeredByRequestFail = false;
-              LOGGER.info("Successfully init store client by async store init thread");
-              break;
-            } catch (Exception ee) {
-              if (ee instanceof InterruptedException || !LatencyUtils.sleep(asyncStoreInitSleepIntervalMs)) {
-                LOGGER.warn("Async store init thread got interrupted, will exit the loop");
+        if (asyncStoreInitThread == null) {
```

Comment:
Looks like in function `close()` at  L719-721 , `asyncStoreInitThread` is also accessed and updated.

```
    if (asyncStoreInitThread != null) {
      asyncStoreInitThread.interrupt();
    }
```

Does it make sense to put move them into  `synchronized` block as well?

Reply:
I am planning to keep the current behavior for now, since today, the store client behavior is non-deterministic after `close` function is called.
I can leave a comment there.

Reply:
I am planning to leave it as it is for now.
Currently, the store client behavior is non-deterministic if the `close` function is being called. 
Having a safeguard in `close` function maybe good for safety, but it will make the logic more complex in other functions and it may hurt performance as well since every get request will need to check whether the store client is closed or not.

I can certainly leave a comment in `close` function.

Reply:
Yeah, let's leave a comment with the this plan.
Just thinking about the worst case, `asyncStoreInitThread` didn't get killed if a race could really happen, and the thread always gets some service discovery error and keeps retrying. Feels maybe that doesn't sounds too bad.